### ISSUE TITLE
[Bug] Error when building `elasticstack_elasticsearch_index_lifecycle` resource

### ISSUE THREAD

**Issue by officialgawney** [elastic/terraform-provider-elasticstack#59]
Hello we're using the elastic provider in terraform to automate the build of some of our elastic resources. 

However when trying to build an ILM policy to our cluster via the **elasticstack_elasticsearch_index_lifecycle** resource the resource plans by fails to apply. 

Specifically we believe it's related the the use of the allocate block - we're able to run a plan / apply successfully without it. 

allocate {
    number_of_replicas = 0
}

**Please see our implementation below followed by the error message we're seeing on running terraform apply** 

**Implementation**

resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default" {
  name = "logs-${module.account_insights.tribe_alias}-default"

  hot {
    min_age = "0ms"
    set_priority {
      priority = 100
    }
    rollover {
      max_primary_shard_size = "10gb"
      max_age = "14d"
    }
  }

  warm {
    min_age = "0ms"
    set_priority {
      priority = 60
    }
    readonly {}
    allocate {
      exclude = "{hot}"
      number_of_replicas = 0
    }
  }

  delete {
    min_age = "31d"
    delete {
      delete_searchable_snapshot = true
    }
    wait_for_snapshot {
      policy = "tvg-s3-policy"
    }
  }
}

**Error message**

Error: Unable to create or update the ILM policy

  with module.elastic_tribe_config.elasticstack_elasticsearch_index_lifecycle.index_lifecycle_default,
  on ../../lifecycle.tf line 1, in resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default":
   1: resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default" {

Failed with:
{"error":{"root_cause":[{"type":"x_content_parse_exception","reason":"Failed
to build [allocate] after last required field
arrived"}],"type":"x_content_parse_exception","reason":"[1:329]
[put_lifecycle_request] failed to parse field
[policy]","caused_by":{"type":"x_content_parse_exception","reason":"[1:329]
[lifecycle_policy] failed to parse field
[phases]","caused_by":{"type":"x_content_parse_exception","reason":"[1:329]
[phases] failed to parse field
[warm]","caused_by":{"type":"x_content_parse_exception","reason":"[1:329]
[phase] failed to parse field
[actions]","caused_by":{"type":"x_content_parse_exception","reason":"[1:329]
[actions] failed to parse field
[allocate]","caused_by":{"type":"x_content_parse_exception","reason":"Failed
to build [allocate] after last required field
arrived","caused_by":{"type":"illegal_argument_exception","reason":"At least
one of include, exclude or requiremust contain attributes for action
allocate"}}}}}}},"status":400}

**Expected behavior**
We'd expect this to plan and apply successfully and the lifecycle to be functional in Elastic.

**Comment by olksdr** [comment]
Hi @officialgawney,

Thanks for opening this issue and reporting a problem you have. I took the example you've posted and tried to run `terraform plan` and immediately getting an error:
```bash
│ Error: "warm.0.allocate.0.exclude" contains an invalid JSON: invalid character 'h' looking for beginning of object key string
│ 
│   with elasticstack_elasticsearch_index_lifecycle.index_lifecycle_default,
│   on resource.tf line 36, in resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default":
│   36:       exclude            = "{hot}"
│ 
╵

```
Which points to the exact place where the problem exists.  The `exclude` field in the `allocate` action must contain a proper JSON object. For example: 

```terraform
allocate {
  exclude = jsonencode({
    box_type = "hot"
  })
  number_of_replicas = 0
}

```

where `box_type` is a custom attribute in the node configuration: `node.attr.box_type: hot` in `elasticsearch.yml`, See more info in: https://www.elastic.co/guide/en/elasticsearch/reference/7.16/ilm-allocate.html#ilm-allocate-assign-index-attribute-ex where you can find an example on how this configured for `include` field, which will work in the same way for `exclude`. 

When I change the definition of the `allocate` action in your config I get the proper plan:
``` 
# elasticstack_elasticsearch_index_lifecycle.index_lifecycle_default will be created
  + resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default" {
      + id            = (known after apply)
      + modified_date = (known after apply)
      + name          = "logs-default"

      + delete {
          + min_age = "31d"

          + delete {
              + delete_searchable_snapshot = true
            }

          + wait_for_snapshot {
              + policy = "tvg-s3-policy"
            }
        }

      + hot {
          + min_age = "0ms"

          + rollover {
              + max_age                = "14d"
              + max_primary_shard_size = "10gb"
            }

          + set_priority {
              + priority = 100
            }
        }

      + warm {
          + min_age = "0ms"

          + allocate {
              + exclude            = jsonencode(
                    {
                      + box_type = "hot"
                    }
                )
              + number_of_replicas = 0
            }

          + readonly {
              + enabled = true
            }

          + set_priority {
              + priority = 60
            }
        }
    }

```

but with current version of the provider it **fails** to apply.
 
Which gave me a clue where the bug could be and how to fix it.  And with a fix in https://github.com/elastic/terraform-provider-elasticstack/pull/60  everything works as expected. I could apply the plan mentioned above,  see `terraform show` output:

```terraform
# elasticstack_elasticsearch_index_lifecycle.index_lifecycle_default:
resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default" {
    id            = "yAc8skMMQbWQ_XXXXXX/logs-default"
    modified_date = "2022-02-04T06:43:12.107Z"
    name          = "logs-default"

    delete {
        min_age = "31d"

        delete {
            delete_searchable_snapshot = true
        }

        wait_for_snapshot {
            policy = "tvg-s3-policy"
        }
    }

    hot {
        min_age = "0ms"

        rollover {
            max_age                = "14d"
            max_docs               = 0
            max_primary_shard_size = "10gb"
        }

        set_priority {
            priority = 100
        }
    }

    warm {
        min_age = "0ms"

        allocate {
            exclude            = jsonencode(
                {
                    box_type = "hot"
                }
            )
            include            = jsonencode({})
            number_of_replicas = 0
            require            = jsonencode({})
        }

        readonly {
            enabled = true
        }

        set_priority {
            priority = 60
        }
    }
}


```

**Comment by officialgawney** [comment]
Hi @olksdr. We've update our config with the required jsonencode (see below) and we're getting a new error message during the plan (also below) . We're still using version 0.2.0 - so presume that we don't have the changes you made as part of #60 .

**Our new implementation**

`resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default" {
  name = "logs-${module.account_insights.tribe_alias}-default"

  hot {
    min_age = "0ms"
    set_priority {
      priority = 100
    }
    rollover {
      max_primary_shard_size = "10gb"
      max_age                = "14d"
    }
  }

  warm {
    min_age = "0ms"

    allocate {
      exclude            = jsonencode({})
      include            = jsonencode({})
      number_of_replicas = 0
      require            = jsonencode({})
    }

    readonly {}

    set_priority {
      priority = 60
    }
  }

  delete {
    min_age = "31d"
    delete {
      delete_searchable_snapshot = true
    }
    wait_for_snapshot {
      policy = "tvg-s3-policy"
    }
  }
}`

**Error message**

`  with module.elastic_tribe_config.elasticstack_elasticsearch_index_lifecycle.index_lifecycle_default,
  on ../../lifecycle.tf line 1, in resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default":
   1: resource "elasticstack_elasticsearch_index_lifecycle" "index_lifecycle_default" {

Failed with:
{"error":{"root_cause":[{"type":"x_content_parse_exception","reason":"[1:339]
[allocate] exclude doesn't support values of type:
VALUE_STRING"}],"type":"x_content_parse_exception","reason":"[1:339]
[put_lifecycle_request] failed to parse field
[policy]","caused_by":{"type":"x_content_parse_exception","reason":"[1:339]
[lifecycle_policy] failed to parse field
[phases]","caused_by":{"type":"x_content_parse_exception","reason":"[1:339]
[phases] failed to parse field
[warm]","caused_by":{"type":"x_content_parse_exception","reason":"[1:339]
[phase] failed to parse field
[actions]","caused_by":{"type":"x_content_parse_exception","reason":"[1:339]
[actions] failed to parse field
[allocate]","caused_by":{"type":"x_content_parse_exception","reason":"[1:339]
[allocate] exclude doesn't support values of type:
VALUE_STRING"}}}}}},"status":400}`


**Provider**

`terraform {
  required_providers {
    elasticstack = {
      source  = "elastic/elasticstack"
      version = "0.2.0"
    }
  }
}`

Many thanks

**Comment by olksdr** [comment]
Thanks for additional info @officialgawney. But the fix is necessary , since the `allocate` action in version `0.2.0` has a bug, which does not properly handle the JSON objects. And the linked PR fixes it. 

For now if you want to use ILM with this provider, you will need to remove `allocate` action from the configuration, which you can add later once the new version is released.

**Comment by officialgawney** [comment]
Thanks for the info @olksdr and your speedy response on this matter. Are you able to give me an ETA on when the new version is going to be released?

Many thanks - Matt

**Comment by olksdr** [comment]
@officialgawney I'm currently working on one more issues, and once it's fixed and merged I will be releasing a new version. I'm looking at the end of this week or the beginning of next one.

**Comment by officialgawney** [comment]
@olksdr thanks for the info.

Matt

**Comment by olksdr** [comment]
New version of the provider released: [v0.3.0](https://github.com/elastic/terraform-provider-elasticstack/releases/tag/v0.3.0)
